Inherit unsafe regions in local functions

agocke opened the pull request:
**Customer scenario**

When local functions are declared in unsafe regions (unsafe statement,
unsafe member, unsafe type), their bodies should be considered unsafe
regions, even if unsafe is not a declaration modifier on the local
function.

**Bugs this fixes:** 

Fixes #13172

**Workarounds, if any**

Explicitly add the `unsafe` modifier to every local function.

**Risk**

Only affects unsafe handling for local functions. Other test cases already handle unsafe code executed in local functions, this simply adds another way to mark the local function unsafe.

**Performance impact**

Low -- an extra parameter in a method call and a single boolean field for every local function symbol.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

It was initially unclear whether or not this would be by design.

**How was the bug found?**

Customer reported.

agocke commented:
/cc @dotnet/roslyn-compiler for review

DustinCampbell commented:
Rebasing to master. We'll switch to the dev15-rc3 branch later -- likely when we enter escrow. Note that all changes in master are in ask mode as well and will require approval to check in.

agocke commented:
test windows_release_unit64_prtest please

agocke commented:
@dotnet/roslyn-compiler Ping for review, please

cston commented on src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs:
```diff
@@ -1091,5 +1091,61 @@ void L2()
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "Foo()").WithArguments("x").WithLocation(7, 13),
+                // (12,31): error CS0165: Use of unassigned local variable 'x'
+                //             System.Action a = Foo;
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "Foo").WithArguments("x").WithLocation(12, 31));
```
Why does assignment of local function value result in unassigned local variable `x` rather than unassigned local `Foo`?

cston replied:
Consider renaming local function to `Local` in this test and test below.

DustinCampbell commented:
@agocke: I'm afraid this picked up a couple of stray commits from dev15-rc3 when it was rebased to master. Could you cherry-pick your PR to get rid of these?

AlekseyTs commented:
@agocke Do you still want to merge this change? Please get rid of the unrelated commits added by the rebase and then ping the team for review.

agocke commented:
@AlekseyTs Done

/cc @dotnet/roslyn-compiler  For review please

cston commented:
LGTM

cston commented:
LGTM

agocke commented:
ping @dotnet/roslyn-compiler For review

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs:
```diff
@@ -306,21 +306,27 @@ public unsafe IEnumerable<int> M4(int* a)
                 // (8,37): error CS1637: Iterators cannot have unsafe parameters or yield types
                 //         IEnumerable<int> Local(int* a) { yield break; }
                 Diagnostic(ErrorCode.ERR_UnsafeIteratorArgType, "a").WithLocation(8, 37),
+                // (8,26): error CS1629: Unsafe code may not appear in iterators
```
>// (8,26): error CS1629: Unsafe code may not appear in iterators [](start = 16, length = 64)

This error feels unexpected.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/LocalScopeBinder.cs:
```diff
@@ -268,7 +268,8 @@ protected LocalFunctionSymbol MakeLocalFunction(LocalFunctionStatementSyntax dec
                 this.ContainingMemberOrLambda,
-                declaration);
+                declaration,
+                this.InUnsafeRegion);
```
The same question applies to this.ContainingMemberOrLambda argument.

---
In reply to: [91799297](https://github.com/dotnet/roslyn/pull/15596#discussion_r91799297) [](ancestors = 91799297)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs:
```diff
@@ -76,6 +77,8 @@ internal ReturnTypeAndDiagnostics(TypeSymbol returnType, ImmutableArray<Diagnost
                 DeclarationModifiers.Static |
                 syntax.Modifiers.ToDeclarationModifiers();
 
+            this.IsUnsafe = inUnsafeRegion || _declarationModifiers.HasFlag(DeclarationModifiers.Unsafe);
```
>_declarationModifiers.HasFlag(DeclarationModifiers.Unsafe) [](start = 46, length = 58)

Can local functions be marked unsafe? It doesn't look like there are any tests for that scenario.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs:
```diff
@@ -76,6 +77,8 @@ internal ReturnTypeAndDiagnostics(TypeSymbol returnType, ImmutableArray<Diagnost
                 syntax.Modifiers.ToDeclarationModifiers();
 
+            this.IsUnsafe = inUnsafeRegion || _declarationModifiers.HasFlag(DeclarationModifiers.Unsafe);
+
```
It looks like this is not the right way to address the problem. In my opinion the source of the problem is a special case for local functions in InMethodBionder.GetFlags. We shouldn't have that special case and we definitely shouldn't erase BinderFlags.UnsafeRegion based on the value of IsUnsafe property. By analogy to the regular methods, this property is supposed to reflect modifiers on the local function itself. When the function is marked unsafe itself the BinderFlags.UnsafeRegion flag should be explicitly injected into the binders chain at the appropriate level, see  how WithUnsafeRegionIfNecessary is used for a regular method.

agocke commented:
@AlekseyTs I think I've addressed your comments. Regarding tests of local functions including the `unsafe` modifier, there are a few of those tests in `CodeGenLocalFunctionTests` where we test both that `unsafe` is respected and also that the unsafe code behaves as expected.

AlekseyTs commented:
LGTM

agocke commented:
@cston I've change my approach here a bit, if you wouldn't mind re-reviewing.

cston commented:
LGTM

agocke commented:
@MattGertz For Ask Mode approval
